### ISSUE TITLE
Add WebSocket transport to Murmur

### ISSUE THREAD

**Issue by mkrautz** [mumble-voip/mumble#2131]
It would be nice if Murmur supported a websocket transport.

This would make it possible to write a Mumble client in the browser, without any intermediary proxies.
(Voice support would be a hack, probably provided via getUserMedia/ScriptProcessorNode, potentially codecs via Emscripten...)

There are a couple of ways to go:

 - Murmur could provide a new listener on port 443, providing WebSocket access. (This would probably clash with existing hosting providers...)
 - Murmur could provide WebSocket service on the ports already used by the regular Mumble protocol. Technically, this would use ALPN and present both HTTP (probably HTTP2 [1]) and the Mumble protocol.
- Both! We could do both, making the port 443 part optional.

[1]: One would assume that when using ALPN, that browsers treat the "default" protocol of a server to be HTTP/1.1. Since we need to be backwards compatible with older Mumble clients, we need the default protocol to be Mumble, not HTTP.
Still, in this scenario, browsers without HTTP2 support that connect to it, will try to send HTTP to Murmur on the listener meant for the Mumble protocol.
It's possible this can be avoided by filtering clients based on their TLS handshakes, for example the ALPN protocol preferences of the client, and other things.

**Comment by eligrey** [comment]
Opus is supported in WebRTC in most browsers iirc.

**Comment by hacst** [comment]
@eligrey It better should be. Opus is a mandatory codec for WebRTC support ;)

**Comment by mkrautz** [comment]
Right, but whether we can use it as-is depends on the API, framing, etc.

**Comment by jsaak** [comment]
I am interested in implementing this. Do you use client to client com? Or everything goes through the server? (the guys at discord did something which works inside a browser) I am still investigating the problem space.

**Comment by hacst** [comment]
@jsaak  It's a classic client-server architecture with a custom protocol. You can take a look at the start of https://mumble-protocol.readthedocs.io/en/latest/ to get the gist of it. Adding websockets would be a completely server-side thing.

**Comment by ubiGG** [comment]
@jsaak This would be an important advancement on the free-software
chessboard, and a very apt interface for our audiences- as long as the
communication is truly full-duplex, and not just a simulation of
full-duplex.

Not only would it be a completely server-side thing, but it almost seems
like architecture might be re-examined, to support many such interfaces.
That's a fairly strong stance on it however, and I'm not sure what other
interfaces would be implemented. I can think of, perhaps, four.

We are interested in collaboration.

**Comment by NickSutton** [comment]
Did this go anywhere? I would like mumble to be able to distribute messages to clients with the username of the person transmitting. Clients are running a GoLang implementation of mumble, interfaced with a html5 page that only requires PTT (by way of touchscreen on Rasp Pi). It would be a nice touch to launch a pop up to display on the other clients who is speaking!

**Comment by bontibon** [comment]
@NickSutton WebSocket supported was added to grumble (mumble-voip/grumble/pull/18). You could try that implementation out and see if it works for you. I am not aware of anyone who is currently implementing this functionality into murmur.

**Comment by porjo** [comment]
Websockets is so 2016 :grinning: Surely WebRTC is the future for a browser based Mumble client? https://github.com/pions/webrtc is making great progress on building a native Go library

**Comment by EPNW** [comment]
WebSocket is something that got me interested recently, and it might be nice to use it with mumble. I'm not planning working on it yet (since I already have an other open mumble PR on my desk...), but share my thoughts. First I want to mention that for the usecase I have in mind (using it with [dumble](https://github.com/EPNW/dumble) as client), using WebSocket as transport (without webRTC) is sufficient, as a way to deal with opus packages is already in place in the dumble infrastructure. As said above by @mkrautz, the existing mumble protocol needs to be the default protocol, and backwards compatibility is needed.

In the WebSocket protocol, the server must only send messages after the client send a handshake. Examing the mumble code I found out that the mumble server sends a version message right after tls is etablished, without waiting for a message from the client. The idea is now to define that the mumble server only sends a version message if it recieved a version message from the client first. I don't know if different mumble clients rely on the server sending the version message first. The dumble client for example just sends its own version message right after a connection is etablished, without waiting for the server, so the propsed approach should not be a problem for this kind of client.

Next, the server needs a way to find out if the client wants to speak "mumble protocol over WebSocket" or "raw mumble protocol". Currently, the mumble protocol requires 2 byte of message type, then 4 byte of message length, followed by the actual message. We could define a new message type `0x71 0x69` which corresponds to the ascii letters `GE` and are the mandatory first letters of a WebSocket request header. If a message of this type is the first message a client sends, the connection is treated as a WebSocket connection. Instead of reading mumble protocol from the socket directly, all data could be passed from the socket to a transport agnostic WebSocket framework (like [WebSocket++](https://github.com/zaphoyd/websocketpp)) containing the WebSocket state machine, and read from there. Then processing could be done as usual. For writing, data would also be passed to the WebSocket framework and written from there instead of writing to the socket directly.

**Comment by EPNW** [comment]
I've implemented the serverside part of websockets here: https://github.com/EPNW/mumble/tree/ws

For more intense testing I'll need to come up with a clientside implementation first.

**Comment by Krzmbrzl** [comment]
With regards to the client certificates: Not sending a certificate is not a good option as we currently tend towards requiring a certificate for all connected clients (that is we are thinking on removing the possibility to connect without a cert). The reason for that being that the cert is used in a lot of places in order to uniquely identify any given user and thus all users that are not using a certificate either require separate handling or for those the respective features simply won't work. And at this point there are qiute a few things that don't work without a cert already.

**Comment by azc5OQ** [comment]
>Murmur could provide a new listener on port 443, providing WebSocket access. (This would probably clash with existing hosting providers...)

WSS port can by almost any port, not just 443
